### ISSUE TITLE
The :profdel command does not seem to have any effect

### ISSUE THREAD

**Issue by dpelle** [vim/vim#3121]
The :profdel command does not seem to have any effect.

Consider the following vim script "profdel-bug.vim":
```
profile start prof.log

func Foo1()
endfunc

func Foo2()
endfunc

func Foo3()
endfunc

profile func Foo1
profile func Foo2

call Foo1()
call Foo2()
call Foo3()

profdel func Foo2

call Foo1()
call Foo2()
call Foo3()
```

Run this script using:
```
$ vim --clean -c 'source profdel-bug.vim' -c q
```

It creates a file prof.log when vim exits with profiling results.
I'd expect prof.log to indicate that:
* Foo1 is called twice
* Foo2 is called once
* Foo3 is not called

But instead it reports that:
* Foo1 is called twice
* Foo2 is called twice
* Foo3 is not called

It is as if the "profdel func Foo2" had no effect. 
Similarly, trying to use "profdel *" instead also has no effect.

It looks like a bug.

I noticed this while trying to add a test for profdel command which is currently not covered by tests according to codecov.

I'm using vim-8.1.127 (huge) on Linux x86_64.

**Comment by brammool** [comment]
Dominique wrote:

> The :profdel command does not seem to have any effect.
> 
> Consider the following vim script "profdel-bug.vim":
> ```
> profile start prof.log
> 
> func Foo1()
> endfunc
> 
> func Foo2()
> endfunc
> 
> func Foo3()
> endfunc
> 
> profile func Foo1
> profile func Foo2
> 
> call Foo1()
> call Foo2()
> call Foo3()
> 
> profdel func Foo2
> 
> call Foo1()
> call Foo2()
> call Foo3()
> ```
> 
> Run this script using:
> ```
> $ vim --clean -c 'source profdel-bug.vim' -c q
> ```
> 
> It creates a file prof.log when vim exits with profiling results.
> I'd expect prof.log to indicate that:
> * Foo1 is called twice
> * Foo2 is called once
> * Foo3 is not called
> 
> But instead it reports that:
> * Foo1 is called twice
> * Foo2 is called twice
> * Foo3 is not called
> 
> It is as if the "profdel func Foo2" had no effect. 
> Similarly, trying to use "profdel *" instead also has no effect.
> 
> It looks like a bug.
> 
> I noticed this while trying to add a test for profdel command which is
> currently not covered by tests according to codecov.

When a function is executed and matches a profile rule, the uf_profiling
flag is set.  When deleting the rule, the uf_profiling flag for matching
functions is not reset.  That is because the flag may also have been set
because it's in a matching file.

Deleting the rule does work before the function is called.  If you
remove the three calls before "profdel" in your example, you'll see that
Foo2() is not profiled.

I'm not sure we need to fix this.  And it's not that easy either.
Just resetting the uf_profiling flag won't work, it would clear the info
stored so far every time.  Hmm, we could add a flag that initialization
was already done.

-- 
hundred-and-one symptoms of being an internet addict:
148. You find it easier to dial-up the National Weather Service
     Weather/your_town/now.html than to simply look out the window.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
This was fixed in vim-8.1.0130 (ad6480961080f80a455b2394f27b02935a2ded52) and tests were added in vim-8.1.0131 (1fbfe7c48cb711f5a6deae535b3ec3bfe7952ce9).

Closing.
